[JENKINS-47043] Introduce exit lifecycle to exit instead of restart

alonbl opened the pull request:
If hudson.lifecycle.ExitLifecycle is selected, jenkins will exit when
restart is requested. Exit code is taken out of
jenkins.model.Jenkins.exitCodeOnRestart, default is 5.

This is usable when there is an external watchdog such as SystemD.

SystemD service example:

   ExecStart=@/usr/bin/java jenkins $JAVA_OPTIONS -DJENKINS_HOME=${JENKINS_HOME} -Dhudson.lifecycle=hudson.lifecycle.ExitLifecycle -Djenkins.model.Jenkins.exitCodeOnRestart=55 -jar $JENKINS_WAR --debug=${JENKINS_DEBUG_LEVEL} --httpPort=${JENKINS_PORT} --httpListenAddress=${JENKINS_LISTEN_ADDRESS} $JENKINS_ARGS

alonbl commented:
@daniel-beck what do you think?

oleg-nenashev commented on core/src/main/java/hudson/lifecycle/UnixLifecycle.java:
```diff
@@ -42,20 +43,25 @@
  * @since 1.304
  */
 public class UnixLifecycle extends Lifecycle {
+    public static String EXIT_ON_RESTART = Configuration.getStringConfigParameter("exitOnRestart", null);
```
Same as in #3035

oleg-nenashev commented on core/src/main/java/hudson/lifecycle/UnixLifecycle.java:
```diff
@@ -98,10 +108,12 @@ public void verifyRestartable() throws RestartNotSupportedException {
-            throw new RestartNotSupportedException("Restart is not supported on Mac OS X");
-        if (args==null)
-            throw new RestartNotSupportedException("Failed to obtain the command line arguments of the process",failedToObtainArgs);
+        if (EXIT_ON_RESTART == null) {
```
unrelated change?

alonbl replied:
why? we do not need these validation if we use this option, right?

oleg-nenashev replied:
sorry, NVM. You are right

oleg-nenashev commented on core/src/main/java/hudson/lifecycle/UnixLifecycle.java:
```diff
@@ -84,10 +90,14 @@ public void restart() throws IOException, InterruptedException {
+            LIBC.execvp(exe, new StringArray(args.toArray(new String[args.size()])));
+            throw new IOException("Failed to exec '"+exe+"' "+LIBC.strerror(Native.getLastError()));
+        } else {
+            System.exit(Integer.parseInt(EXIT_ON_RESTART));
```
The field should be `Integer` then. No need to parse it in runtime and to risk that the logic blows up here in the case of typo in the argument

alonbl replied:
ok

alonbl commented:
ping?

oleg-nenashev commented:
@alonbl GitHub does not send mail notifications if you force to the branch. Will review

olivergondza commented:
I would be much cleaner to introduce new `Lifecycle` that would exit on restart and let all these watchdog programs to request it explicitly via existing `hudson.lifecycle` property.

alonbl commented:
ok, will try.

alonbl committed:
[JENKINS-47043] Introduce exit lifecycle to exit instead of restart

If hudson.lifecycle.ExitLifecycle is selected, jenkins will exit when
restart is requested. Exit code is taken out of
jenkins.model.Jenkins.exitCodeOnRestart, default is 5.

This is usable when there is an external watchdog such as SystemD.

SystemD service example:

   ExecStart=@/usr/bin/java jenkins $JAVA_OPTIONS -DJENKINS_HOME=${JENKINS_HOME} -Dhudson.lifecycle=hudson.lifecycle.ExitLifecycle -Djenkins.model.Jenkins.exitCodeOnRestart=55 -jar $JENKINS_WAR --debug=${JENKINS_DEBUG_LEVEL} --httpPort=${JENKINS_PORT} --httpListenAddress=${JENKINS_LISTEN_ADDRESS} $JENKINS_ARGS

Signed-off-by: Alon Bar-Lev <alon.barlev@gmail.com>

alonbl commented:
@oleg-nenashev : done

olivergondza commented on core/src/main/java/hudson/lifecycle/ExitLifecycle.java:
```diff
@@ -0,0 +1,66 @@
+import jenkins.model.Jenkins;
+
+/**
+ * {@link Lifecycle} implementation when Hudson runs on the embedded
```
What's "Hudson"?

alonbl replied:
I copied this from here[1], do you prefer I use Jenkins?

[1] https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/lifecycle/UnixLifecycle.java#L40

olivergondza replied:
"Lifecycle that delegates the responsibility to restart Jenkins to an external entity such as starting process or a service management of some sort."?

alonbl replied:
Done

oleg-nenashev commented on core/src/main/java/hudson/lifecycle/ExitLifecycle.java:
```diff
@@ -0,0 +1,66 @@
+/*
+ * The MIT License
```
Not valid without copyright holder IIRC. IANAL though

alonbl replied:
ok, added self.

oleg-nenashev commented on core/src/main/java/hudson/lifecycle/ExitLifecycle.java:
```diff
@@ -0,0 +1,66 @@
+        System.exit(exitOnRestart);
+    }
+
+    private static final Logger LOGGER = Logger.getLogger(ExitLifecycle.class.getName());
```
it would be better to keep the logger field in the top. It's what the Oracle codestyle recommend though not everybody follows it in the core

alonbl replied:
ok, I copied this from the UnixLifecycle...
Done.

oleg-nenashev commented on core/src/main/java/hudson/lifecycle/ExitLifecycle.java:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @author Alon Bar-Lev
+ */
+public class ExitLifecycle extends Lifecycle {
```
There is no `@Extension` annotation. Should be functionally fine, but in such case the extension point won't be posted in https://jenkins.io/doc/developer/extensions/jenkins-core/#lifecycle

alonbl replied:
Done, do you want me to submit another patch to fix the other implementations?

oleg-nenashev commented on core/src/main/java/hudson/lifecycle/ExitLifecycle.java:
```diff
@@ -0,0 +1,66 @@
+ * Restart by exit with specific code.
+ *
+ * @author Alon Bar-Lev
+ */
```
Please add `@Restricted(NoExternalUse.class)` so that we will be able to adjust the location/API later if needed

alonbl replied:
Done, do you want me to submit another patch to fix the other implementations?

oleg-nenashev replied:
would be useful, but it should be in another PR

oleg-nenashev reviewed (commented):
The code generally looks good to me, but I have a concern about the approach (**UPD:** to be explicit, I am fine with the current state if the class is restricted)

* The current lifecycle replaces standard ones if enabled (e. .g. https://github.com/jenkinsci/jenkins/blob/4ab693846ca7a4aa112959a2e92688c3fb9122c3/core/src/main/java/hudson/lifecycle/UnixLifecycle.java)
* It implements only the restart method and suppresses other behaviors
* This installer patch may be useful together with other lifecycles

Maybe it would be better to introduce a new `LifecycleAppender` which would be able to modify existing lifecycles. But maybe it is an overkill.

olivergondza commented on core/src/main/java/hudson/lifecycle/ExitLifecycle.java:
```diff
@@ -0,0 +1,66 @@
+public class ExitLifecycle extends Lifecycle {
+
+    private static final String EXIT_CODE_ON_RESTART = "exitCodeOnRestart";
+    private static final String DEFAULT_EXIT_CODE = "5";
```
Is there any use-case for specifying the exit code used? I would argue it should be 0, by default at least.

alonbl replied:
We must provide a way to distinguish between stop and restart. Stop is exit code 0, restart should be something else, I use 5 as default it can be anything, I can put 55 as in the commit message to emphasis it as unique.

Setting it to 0 will result in stop instead of restart, maybe this what you want.

oleg-nenashev reviewed (approved):
LGTM. @olivergondza are you fine with the PR and the explanation of the default exit code? IMHO whatever works while it is configurable and documented

daniel-beck commented:
@oleg-nenashev 

> It implements only the restart method and suppresses other behaviors

Seems complete to me, UnixLifecycle does the same thing.

oleg-nenashev commented:
Agreed. I think it is ready to go. @daniel-beck I put it on hold so you have time to review it.

daniel-beck commented:
🤷‍♂️ Assuming this works and won't need followup options for tons of different things, sure. I'm not familiar enough with the problem domain to say one way or the other. Perhaps @rtyler or @olblak can chime in.

Would it be useful to have a linked PR in the Docker image or native installers to apply this option?

oleg-nenashev commented:
> Would it be useful to have a linked PR in the Docker image or native installers to apply this option?

It would require a more careful review by maintainers. IIUC https://github.com/jenkinsci/docker does not behave in a compatible way with proper failover. It would require external monitors/failover logic in Kubernetes, Docker Swarm/Compose or whatever

daniel-beck commented:
@oleg-nenashev So how do we know this feature as implemented actually makes sense?

alonbl commented:
Hi,

Per systemd, it is fully supported with:
```
Restart=on-success
SuccessExitStatus=5
RestartPreventExitStatus=0
```

Docker wrapper script can be:
```
while true; do
    java....
    status=$?
    [ $status -eq 5 ] || exit $status
done
```

k8s is bit behind, but can enjoy the same as docker.
https://github.com/kubernetes/kubernetes/issues/31147

Regards,
Alon
